### DISCUSSION TITLE
convergence problems

### DISCUSSION THREAD

**Discussion by asgauthier** [idaholab/moose#24967] [Q&A Modules: General]
Hello everyone,

I'm asking you for help because I did a 3D model with faults and a HydroThermoMechanical coupling with PorousFlowFullySaturated in my input file but I do not manage to make it converge. I don't know if it is a problem of scaling, or the mesh, or the parameters so if you have any idea tell me :D Here is the link to the mesh file https://we.tl/t-DSWRd63Jks
`[Mesh] 
    [basic_mesh]
        type = FileMeshGenerator
        file=Mol_Mesh3.e
    [ ]
    [renamed_mesh]
        type = RenameBlockGenerator
        input = basic_mesh
        old_block = '0 1 2 3'
        new_block = 'matrix Ber Des Ret'
    []  
    [finalMesh]
        type = RenameBoundaryGenerator
        input = renamed_mesh
        old_boundary = '1 2 3 4 5 6 7 8 9'
        new_boundary = 'bottom top Ber Des Ret front left back right'
    []
[ ]
[GlobalParams]#Every parameter in the block will be inserted into every block/sub-block where that parameter name is defined
    PorousFlowDictator = dictator
    displacements = 'disp_x disp_y disp_z'
    biot_coefficient=0.85
[ ]

[Variables]#define the unknowns within a system of partial differential equations
    [porepressure]
        initial_condition=38.3E6
    [ ]
    [temperature]
        initial_condition=401.15 
        scaling=1E-6
    []
    [disp_x]
        scaling=1E-10
    [ ]
    [disp_y]
        scaling=1E-10
    [ ]
    [disp_z]
        scaling=1E-10
    [ ]   
[ ]

[BCs]
    [porepressure]
        type = DirichletBC #Imposes the essential boundary condition u=g
        value = 38.3E6 #pression initiale du fluide en Pascal
        variable = porepressure
        boundary = 'right left top bottom back front'
    [ ]
    [temperature]
        type = DirichletBC
        variable = temperature
        value = 401.15
        boundary = 'right left top bottom back front'
    []
    [stress_x_right]
        type = NeumannBC #Imposes the integrated boundary condition du/dn=h
        value = -55.75E6 #pourquoi ??
        variable = disp_x
        boundary = right
    [ ]
    [stress_x_left]
        type = NeumannBC
        value = 55.75E6
        variable = disp_x
        boundary = left
    [ ]
    [stress_y_back]
        type = NeumannBC
        value = -55.75E6
        variable = disp_y
        boundary = back
    [ ]
    [stress_y_front]
        type = NeumannBC
        value = 55.75E6
        variable = disp_y
        boundary = front
    [ ]
    [stress_z_top]
        type = NeumannBC
        value = -55.75E6
        variable = disp_z
        boundary = top
    [ ]
    [stress_z_bottom]
        type = NeumannBC
        value = 55.75E6
        variable = disp_z
        boundary = bottom
    [ ]    
[ ]
[PorousFlowFullySaturated]
  porepressure = porepressure
  temperature = temperature
  coupling_type = ThermoHydroMechanical
  gravity = '0 0 0'
  fp = the_simple_fluid
  stabilization=KT
  eigenstrain_names = 'thermal_contribution ini_stress'
  add_stress_aux=true
  use_displaced_mesh = false
[]

[DiracKernels]
  [fluid_injection]
    type = PorousFlowPointSourceFromPostprocessor
    point = '202340 213597 -3126.2'
    mass_flux = 18.05
    variable = porepressure
  []
  [source_h]
    type = PorousFlowPointEnthalpySourceFromPostprocessor
    variable = temperature
    mass_flux = 18.05
    point = '202340 213597 -3126.2'
    T_in = 333.15
    pressure = porepressure
    fp = the_simple_fluid
  []
  [fluid_production]
    type = PorousFlowPointSourceFromPostprocessor
    point = '202340 212731 -3168.34'
    mass_flux = -18.05
    variable = porepressure
  []
[]
[FluidProperties]
    [the_simple_fluid]#Fluid properties for a simple fluid with a constant bulk density
        type = SimpleFluidProperties
        bulk_modulus = 2E9
        viscosity = 1.0E-3
        density0 = 1000.0
        thermal_expansion = 0.0002
        cp = 4194
        cv = 4186
        porepressure_coefficient = 0
    [ ]
[ ]

[Materials]
    [porosity_reservoir]
        type = PorousFlowPorosity
        porosity_zero = 0.024
        mechanical = false 
        fluid = true
        reference_porepressure = 38.3E6
        solid_bulk = 15E9
        block='matrix'
    [ ]
    [porosity_faults]
        type = PorousFlowPorosityConst
        porosity=1
        block='Ber Des Ret'
    [ ]
    [permeability_reservoir]
        type = PorousFlowPermeabilityKozenyCarman #calculates the permeability tensor
        poroperm_function = kozeny_carman_phi0
        k0 = 1.5E-12
        m = 2 #(1-porosity) exponent
        n = 3 #Porosity exponent
        phi0 = 0.024
        block='matrix'
    [ ]
    [permeability_faults]
        type = PorousFlowPermeabilityConst
        permeability='1E-14 0 0 0 1E-14 0 0 0 1E-14'
        block='Ber Des Ret'        
    [ ]
    [rock_internal_energy_rock]
        type = PorousFlowMatrixInternalEnergy
        density = 2700.0
        specific_heat_capacity = 1200.0
        block='matrix'
    []
    [rock_internal_energy_fault]
        type = PorousFlowMatrixInternalEnergy
        density = 2700.0
        specific_heat_capacity = 0
        block='Ber Des Ret'
    []
    [thermal_conductivity]
        type = PorousFlowThermalConductivityIdeal
        dry_thermal_conductivity = '2.6 0 0  0 2.6 0  0 0 2.6'
    []
    [elasticity_tensor]
        type = ComputeIsotropicElasticityTensor
        bulk_modulus = 15E9
        poissons_ratio = 0.281
    []
    [strain]
        type = ComputeSmallStrain
        displacements = 'disp_x disp_y disp_z'
        eigenstrain_names ='thermal_contribution ini_stress'
    [ ]
    [thermal_contribution]
        type = ComputeThermalExpansionEigenstrain
        temperature = temperature
        thermal_expansion_coeff = 10E-5 # this is the linear thermal expansion coefficient
        eigenstrain_name = thermal_contribution
        stress_free_temperature = 401.15
    []
    [stress]
        type = ComputeLinearElasticStress
    [ ]
    [ini_stress]
        type = ComputeEigenstrainFromInitialStress
        initial_stress = ' -55.75E6 0 0 0 -55.75E6 0 0 0 -77.7E6'
        eigenstrain_name = ini_stress
    [ ]
[ ]
    [Postprocessors]#object that computes a single scalar
        [runtime]
            type = PerfGraphData
            data_type = TOTAL
            section_name = 'Root'
        [ ]

        [porepressure]
            type = PointValue
            point = '203457 214307 -3531.87'
            variable = porepressure
        [ ]
        [temperature]
            type = PointValue
            point = '203457 214307 -3531.87'
            variable = temperature
        [ ]

[ ]
[Preconditioning]#to define which type of preconditioning matrix to build and what process to apply
    active = basic
    [basic]
        type = SMP#Single matrix preconditioner (SMP) builds a preconditioner using user defined off-diagonal parts of the Jacobian
        full = true
    [ ]
[ ]
[Executioner]
    type = Transient #Executioner for time varying simulations
    solve_type = Newton
    nl_abs_tol = 1E-3
    end_time = 50
        [TimeStepper]
            type = TimeSequenceStepper
            time_sequence = '0 10 20 30 40 50'
        [ ]
    steady_state_detection = true
[ ]
[Outputs]
    perf_graph = true
    exodus = true
    csv = true
[ ]`

**Comment by GiudGiud** [comment]
Hello

your nonlinear tolerance is likely too loose for starters. You could turn on automatic_scaling in the executioner to get a more fair convergence criterion. 

Then two things: you ll only want to use Newton s method if your Jacobian has been coded 100% correctly. You can check that using the analyze_jacobian utility or by checking if a finite different preconditioner makes a difference. 

Then you ll want to work on preconditioning. You are getting the default right now. If you can afford it you could try a direct solve with LU preconditioning.

All this advice and much more is summarized on the « troubleshooting failed solve » page on the website

Guillaume

**Reply by cpgr** [reply]
The Jacobian should be ok, but I agree the tolerance should be tighter.

As a guess, I would be almost certain that your BC's on displacements are the issue. Could you replace them with roller BC's to test - so DirichletBC's with value = 0 for each displacement?

**Reply by asgauthier** [reply]
This converges better with Dirichlet indeed :D Is there a better way to impose the initial stress state of the rock ?
